src/util/recentlyEditedTree.js

```diff
@@ -0,0 +1,286 @@
+        leafNodes.forEach(descendant => {
+          const descendantContext = pathToContext(descendant.path)
+          if (descendantContext[0] === EM_TOKEN) return // preventing nodes at level 0 from merged to this (temporary fix)
+          const [shortContext, longContext] = newContext.length < descendantContext.length
+            ? [newContext, descendantContext]
+            : [descendantContext, newContext]
+
+          const isSameDepth = descendantContext.length === newContext.length
+          const hasSameParent = newContext.length - commonPath.length === 1
+          const isSibling = isSameDepth && hasSameParent
+
+          /**
+           * if shorter context length and commonPath length differs by more than 1 then then it shouldn't be merged.
+           * only merge when it has distant relation i.e A.B.C and A.B.D.E.F.G.H ---> A.B.D.E.F.G.H
+           * */
+          const notMergeableDepthDiff = shortContext.length - commonPath.length > 1
+          // for restricting merge of direct cousins like A.B.C.F and A.B.D.E
+          const isDirectCousin = newContext.length === descendantContext.length && newContext.length - commonPath.length === 2
+
+          // siblings A.B.C and A.B.D ---> A.B.D
+          if (isSibling) {
+            isMerged = true
+            _.unset(tree, descendantContext)
+            _.set(tree, newContext, { leaf: true, lastUpdated: timestamp(), path: newPath })
+          }
+          /*
+            restricting merge of direct cousins like A.B.C.F and A.B.D.E
+            but merging distant relation A.B.C and A.B.D.E.F.G.H ---> A.B.D.E.F.G.H
+          */
+          else if (!(notMergeableDepthDiff || isDirectCousin)) {
+            isMerged = true
+            _.unset(tree, shortContext.slice(0, commonPath.length + 1))
+            _.set(tree, longContext, { leaf: true, lastUpdated: timestamp(), path: newContext.length > descendantContext.length ? newPath : descendant.path })
+          }
+        })
+        // adding new thought to the tree
+        if (!isMerged) _.set(tree, newContext, { leaf: true, lastUpdated: timestamp(), path: newPath })
+      }
+    }
+  }
+}
+
+/**
+   * Deletes node from existing tree object using mutation.
+   * @param {object} tree nested object representing tree
+   * @param {string[]} oldPath array of string representing path
+   * @param {string[]} newPath array of string representing path
+   * @param {boolean} [timestampUpdate=true] if false it doesn't update lastUpdated property of all affected leaf nodes (used by nodeMove)
+   */
+const nodeDelete = (tree, oldPath, timestampUpdate = true) => {
+  const oldContext = pathToContext(oldPath)
+  const { node: commonNode, path: commonPath } = findTreeDeepestSubcontext(tree, oldContext)
+  if (commonNode) {
+    const isNodeNotAvailable = oldContext.length > commonPath.length
+    if (isNodeNotAvailable) return // if node is not already available in the tree
```

Comment:
Here I would simply inline the conditional since it is unitary. Intermediate variables are justified only when you have a complex composite conditional, such as `if (A && (B || C))` or in some cases just `if (A && B)` if the expressions are long enough.

Reply:
Agreed!